### PR TITLE
stats: Make an object for the FIlesystem (Filesystem::Instance) to hold the stats counters.

### PR DESCRIPTION
**Pull request by jmarantz**
*Description*: Currently, stats counters/gauges are looked up every time a FileImpl is created. This change factors that out, introducing a Filesystem::Instance to hold the stats objects.

This is important for SymbolTables because lookups from string-literals take a lock, and we don't want to take locks while processing requests.
*Risk Level*: low
*Testing*:  //test/...
*Docs Changes*:
*Release Notes*:

### CODE REVIEW

**Comment by jmarantz** [comment]
I think this is failing CI because of an API change to RawConnectionDriver, which now requires a Stats::Store, and this requires changing echo2_integration_test.cc in https://github.com/envoyproxy/envoy-filter-example .  I guess that to avoid breaking any build at any time I have to issue two small PRs first:

 1. change RawConnectionDriver to take an optional stat_store object and ignore it
 2. change echo2_integration_test.cc to pass in a Stat::Store which it can get from BaseIntegrationTest via test_server_.stats().

A variation on this is to have BaseIntegrationTest provide a method for building a RawConnectionDriverPtr -- so that it can provide any other context required in the future. That feels a little nicer because I think it means doing this two-repo-PR dance a little less in the future, should more context be required. My concern is that this might be going against the design goals for RawConnectionDriver.

@mattklein123 @htuch @alyssawilk WDYT?

**Comment by mattklein123** [comment]
@jmarantz at a high level, we don't create files in any high performance path. Is it worth the code churn here? (I haven't actually reviewed yet but trying to understand if the overall change is worth it.)

**Comment by htuch** [comment]
On the envoy-filter-example front, I have an idea how to make this dance less insane:
1. We move envoy-filter-example into the main Envoy repo, using a local repository in the WORKSPACE to point the example at Envoy source.
2. We export to envoy-filter-example as a read-only mirror (like data-plane-api), and swizzle the WORKSPACE to one that uses a remote repository.

This preserves envoy-filter-example as a standalone example of integrating Envoy, while avoiding the multi-phase commit dance (and breakages) we hit regularly these days. WDYT?

**Comment by mattklein123** [comment]
> This preserves envoy-filter-example as a standalone example of integrating Envoy, while avoiding the multi-phase commit dance (and breakages) we hit regularly these days. WDYT?

This sounds awesome. We could also then test it as part of the main CI.

**Comment by jmarantz** [comment]
It's possible this might not be performance critical in real examples, but I also have refactored away for the need to change the API for filter-examples, so I'm hopeful that asan/tsan CI will just pass now.

It might still not be worth the churn, but there's less churn now :)

The reason this crossed my radar is that envoy-perf/siege/siege.py (which I wrote) uses a static file for serving origin content for the load test. IMO it's a justifiable refactor to give the file-system some context.

**Comment by mattklein123** [comment]
OK can you merge master and we can take a look and discuss?

**Comment by htuch** [comment]
> > This preserves envoy-filter-example as a standalone example of integrating Envoy, while avoiding the multi-phase commit dance (and breakages) we hit regularly these days. WDYT?
> 
> This sounds awesome. We could also then test it as part of the main CI.

I've created https://github.com/envoyproxy/envoy/issues/5054 to track. I might have cycles to work on this coming up, so tentatively assigned to me.

**Comment by jmarantz** [comment]
ptal; this is now merged and ready for review.

File: source/common/filesystem/filesystem_impl.h [hunk 1]
```diff
@@ -32,6 +32,44 @@ struct FileSystemStats {
...
+   * @param dispatcher The dispatcher used for set up timers to run flush().
+   * @param lock The lock.
+   */
+  FileSharedPtr createFile(const std::string& path, Event::Dispatcher& dispatcher,
+                           Thread::BasicLockable& lock) {
+    return createFile(path, dispatcher, lock, file_flush_interval_msec_);
+  }
+
+private:
+  std::chrono::milliseconds file_flush_interval_msec_;
```
Comment (mattklein123) [inline source/common/filesystem/filesystem_impl.h]: nit: const

File: source/common/stats/thread_local_store.cc [hunk 2]
```diff
@@ -34,6 +34,48 @@ ThreadLocalStoreImpl::~ThreadLocalStoreImpl() {
   ASSERT(scopes_.empty());
 }
 
+void ThreadLocalStoreImpl::setStatsMatcher(StatsMatcherPtr&& stats_matcher) {
```
Comment (mattklein123) [inline source/common/stats/thread_local_store.cc]: This seems unrelated to filesystems. Thoughts on moving this part of the change to a different PR?

Comment (jmarantz) [reply source/common/stats/thread_local_store.cc]: The reason I needed to tackle this was that pre-declaring the stats for the filesystem at API creation time made them all leak through the late application of the StatsMatcher and broke the StatsMatcher integration test, which already had one hardcoded exception in it.

I felt it better to clean up the matcher than had a lot more exceptions to the integration-test.

However what I can do is put that into its own PR which we can review first.

Comment (mattklein123) [reply source/common/stats/thread_local_store.cc]: Great, SGTM. Thanks for breaking things apart.

Comment (jmarantz) [reply source/common/stats/thread_local_store.cc]: Pulled into #5015 which can be reviewed without the file-system change.

Comment (jmarantz) [reply source/common/stats/thread_local_store.cc]: #5105 is merged now and so is no longer part of this PR.
